### PR TITLE
Improve the retry of the rpc client

### PR DESCRIPTION
**Pull request by SimFG**
issue: #26793

### CODE REVIEW

File: internal/util/grpcclient/client.go [hunk 1]
```diff
@@ -311,55 +314,86 @@ func (c *ClientBase[T]) connect(ctx context.Context) error {
...
+			// it will directly return the result
+			log.Warn("unknown return type", zap.Any("return", ret))
+			return nil
+		}
+
+		if errorCode == commonpb.ErrorCode_Success {
+			return nil
+		}
+
+		if IsCrossClusterRoutingErr(reason) || IsServerIDMismatchErr(reason) {
```
Comment (bigsheeper) [inline internal/util/grpcclient/client.go:378]: CrossClusterRoutingErr and ServerIDMismatchErr come from error, not status.Reason

File: internal/util/grpcclient/client.go [hunk 2]
```diff
@@ -311,55 +314,86 @@ func (c *ClientBase[T]) connect(ctx context.Context) error {
...
-			if coordSess, exist := sessions[c.GetRole()]; exist {
-				if c.GetNodeID() != coordSess.ServerID {
-					log.Warn("Server ID mismatch, may connected to a old server, start to reset connection", zap.Error(err))
-					c.resetConnection(client)
-					return ret, err
-				}
+
+	innerCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	_ = retry.Do(innerCtx, func() error {
```
Comment (bigsheeper) [inline internal/util/grpcclient/client.go:341]: why ignore retriy error

File: internal/util/grpcclient/client.go [hunk 3]
```diff
@@ -383,30 +410,8 @@ func (c *ClientBase[T]) Call(ctx context.Context, caller func(client T) (any, er
...
-		traceErr = fmt.Errorf("err: %w\n, %s", err, tracer.StackTrace())
-		log.Warn("ClientBase ReCall grpc second call get error", zap.Error(traceErr))
-		return generic.Zero[T](), traceErr
-	}
-	return ret, err
+	// All retry operations are done in `call` function.
+	return c.Call(ctx, caller)
 }
 
 func (c *ClientBase[T]) bgHealthCheck(client T) {
```
Comment (wayblink) [inline internal/util/grpcclient/client.go:417]: bgHealthCheck can be removed as no more use

**Review by wayblink — APPROVED** [review]
lgtm with one minor comment

**Comment by SimFG** [comment]
rerun ut

File: internal/util/grpcclient/client.go [hunk 4]
```diff
@@ -311,55 +295,85 @@ func (c *ClientBase[T]) connect(ctx context.Context) error {
...
+			callErr = errors.Wrap(clientErr, "empty grpc client")
+			log.Warn("grpc client is nil, maybe fail to get client in the retry state")
+			resetClientFunc()
+			return callErr
+		}
+		ret, callErr = caller(client)
+		if callErr != nil {
+			if funcutil.IsGrpcErr(callErr) ||
+				IsCrossClusterRoutingErr(callErr) || IsServerIDMismatchErr(callErr) ||
+				!funcutil.CheckCtxValid(ctx) {
```
Comment (czs007) [inline internal/util/grpcclient/client.go:333]: if ctx is canceled or ctx is timeout , we should not resetClientFunc

**Pull request merged**